compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp

```diff
@@ -0,0 +1,602 @@
+    rewriter.finalizeRootUpdate(genericOp);
+    return success();
+  }
+};
+
+// Sinks a transpose to the input of a linalg named op. The conditions for the
+// rewrite are
+//   1) One of the input producers to the named op is a linalg.transpose
+//   2) The named op is generalizable (and is not a transpose)
+// The easiest way to get the rewrite we want then is to just try to generalize
+// all transposed named ops and let the generic pattern handle the actual
+// rewrite.
+class GeneralizeInputTransposedNamedOp
+    : public OpInterfaceRewritePattern<linalg::LinalgOp> {
+public:
+  using OpInterfaceRewritePattern<linalg::LinalgOp>::OpInterfaceRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::LinalgOp linalgOp,
+                                PatternRewriter &rewriter) const override {
+    if (!IREE::Flow::isNonNullAndOutsideDispatch(linalgOp)) {
+      return failure();
+    }
+    // Don't generalize transposes.
+    if (isa<linalg::TransposeOp>(linalgOp)) {
+      return failure();
+    }
+    bool hasTranspose = false;
+    for (Value input : linalgOp.getDpsInputs()) {
+      if (input.getDefiningOp<linalg::TransposeOp>()) {
+        hasTranspose = true;
+        break;
+      }
+    }
+    if (!hasTranspose) {
+      return failure();
+    }
+    if (failed(linalg::generalizeNamedOp(rewriter, linalgOp))) {
+      return failure();
+    }
+    return success();
+  }
+};
+
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// Linalg Named Op -> Named Op Conversions
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+template <typename OpTy, typename ReplTy, int64_t inputIdx>
+class NamedOpConversion : public OpRewritePattern<OpTy> {
+public:
+  using OpRewritePattern<OpTy>::OpRewritePattern;
+  NamedOpConversion(MLIRContext *ctx, SmallVector<int64_t> perm,
+                    PatternBenefit b = 1)
+      : OpRewritePattern<OpTy>(ctx, b), permutation(perm) {}
+
+  LogicalResult matchAndRewrite(OpTy namedOp,
+                                PatternRewriter &rewriter) const override {
+    if (!IREE::Flow::isNonNullAndOutsideDispatch(namedOp)) {
+      return failure();
+    }
+
+    Value input = namedOp.getInputs()[inputIdx];
+    auto transpose = input.getDefiningOp<linalg::TransposeOp>();
+    if (!transpose) {
+      return failure();
+    }
+
+    SmallVector<int64_t> transPerm(transpose.getPermutation());
+    if (transPerm != permutation) {
+      return failure();
+    }
+    SmallVector<NamedAttribute> attrs = getPrunedAttributeList(namedOp);
+    SmallVector<Value> newInputs = namedOp.getInputs();
+    newInputs[inputIdx] = transpose.getInput();
+    rewriter.replaceOpWithNewOp<ReplTy>(namedOp, newInputs,
+                                        namedOp.getDpsInits(), attrs);
+    return success();
+  }
+
+private:
+  // non-type literal array template parameters are a C++20 feature.
```

Comment:
remove the outdated comment?

Reply:
I clarified it. I basically wanted to explain why the permutation wasn't passed as a template argument.